atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java

```diff
@@ -75,6 +75,10 @@ public long leaderPingResponseWaitMs() {
 
     @Value.Check
     protected final void check() {
+        Preconditions.checkState(ImmutableSet.copyOf(leaders()).size() == leaders().size(),
```

Comment:
I'm still semi-opposed to the switch from set to list, and to address the concerns about potential duplicate entries I feel like these checks could use some more love to be more explicit about misconfiguration. These ensure that we have at least enough unique leaders to get desired quorum and also validate that desired quorum is correct for the number of potential leaders.


```
        Preconditions.checkState(quorumSize() >= minimumQuorumSize(),
                "The desired quorum size %s must be greater than or equal to minimum acceptable quorum size %s for %s leaders: %s",
                quorumSize(), minimumQuorumSize(), leaders().size(), leaders());

        Preconditions.checkState(leaders().size() >= minimumQuorumSize(),
                "There must be at least %s leaders to reach minimum quorum of %s, found %s leaders: %s",
                minimumQuorumSize(), quorumSize(), leaders().size(), leaders());

        Preconditions.checkState(leaders().size() >= quorumSize(),
                "There must be at least %s leaders to reach desired quorum of %s (minimum acceptable quorum %s), found %s leaders: %s",
                quorumSize(), quorumSize(), minimumQuorumSize(), leaders().size(), leaders());
```

and corresponding tests to verify the behavior and messages:

```
    @Test
    public void quorumSizes() throws Exception {
        assertThat(LeaderConfig.minimumQuorumSize(1), is(1));
        assertThat(LeaderConfig.minimumQuorumSize(2), is(2));
        assertThat(LeaderConfig.minimumQuorumSize(3), is(2));
        assertThat(LeaderConfig.minimumQuorumSize(4), is(3));
        assertThat(LeaderConfig.minimumQuorumSize(5), is(3));
        assertThat(LeaderConfig.minimumQuorumSize(6), is(4));
        assertThat(LeaderConfig.minimumQuorumSize(7), is(4));
        assertThat(LeaderConfig.minimumQuorumSize(8), is(5));
        assertThat(LeaderConfig.minimumQuorumSize(9), is(5));
        assertThat(LeaderConfig.minimumQuorumSize(10), is(6));
        assertThat(LeaderConfig.minimumQuorumSize(11), is(6));
        assertThat(LeaderConfig.minimumQuorumSize(12), is(7));
        assertThat(LeaderConfig.minimumQuorumSize(13), is(7));
    }

    @Test
    public void oneLeaderQuorumOne() {
        ImmutableLeaderConfig.builder()
                .localServer("a")
                .addLeaders("a")
                .quorumSize(1)
                .build();
    }

    @Test
    public void twoLeadersQuorumOne() {
        try {
            ImmutableLeaderConfig.builder()
                    .localServer("a")
                    .addLeaders("a", "b")
                    .quorumSize(1)
                    .build();
            fail("Should have thrown");
        } catch (IllegalStateException expected) {
            assertThat(expected.getMessage(),
                    is("The desired quorum size 1 must be greater than or equal to minimum acceptable quorum size 2 for 2 leaders: [a, b]"));
        }
    }

    @Test
    public void twoLeadersQuorumTwo() {
        ImmutableLeaderConfig.builder()
                .localServer("a")
                .addLeaders("a", "b")
                .quorumSize(2)
                .build();
    }

    @Test
    public void twoLeadersQuorumThree() {
        try {
            ImmutableLeaderConfig.builder()
                    .localServer("a")
                    .addLeaders("a", "b")
                    .quorumSize(3)
                    .build();
            fail("Should have thrown");
        } catch (IllegalStateException expected) {
            assertThat(expected.getMessage(),
                    is("There must be at least 3 leaders to reach desired quorum of 3 (minimum acceptable quorum 2), found 2 leaders: [a, b]"));
        }
    }

    @Test
    public void threeLeadersQuorumOne() {
        try {
            ImmutableLeaderConfig.builder()
                    .localServer("a")
                    .addLeaders("a", "b", "c")
                    .quorumSize(1)
                    .build();
            fail("Should have thrown");
        } catch (IllegalStateException expected) {
            assertThat(expected.getMessage(),
                    is("The desired quorum size 1 must be greater than or equal to minimum acceptable quorum size 2 for 3 leaders: [a, b, c]"));
        }
    }

    @Test
    public void threeLeadersQuorumTwo() {
        ImmutableLeaderConfig.builder()
                .localServer("a")
                .addLeaders("a", "b", "c")
                .quorumSize(2)
                .build();
    }

    @Test
    public void threeLeadersQuorumThree() {
        ImmutableLeaderConfig.builder()
                .localServer("a")
                .addLeaders("a", "b", "c")
                .quorumSize(3)
                .build();
    }

    @Test
    public void threeLeadersQuorumFour() {
        try {
            ImmutableLeaderConfig.builder()
                    .localServer("a")
                    .addLeaders("a", "b", "c")
                    .quorumSize(4)
                    .build();
            fail("Should have thrown");
        } catch (IllegalStateException expected) {
            assertThat(expected.getMessage(),
                    is("There must be at least 4 leaders to reach desired quorum of 4 (minimum acceptable quorum 2), found 3 leaders: [a, b, c]"));
        }
    }
```